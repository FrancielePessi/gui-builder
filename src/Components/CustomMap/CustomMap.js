import React, { useEffect, useRef } from 'react';
import L from 'leaflet';
import PropTypes from 'prop-types';
import Util from '../../utils';


const CustomMap = ({ markersData, center, zoom }) => {
    // create map
    const sid = Util.sid();
    const mapRef = useRef(sid);
    const xCenter = center || [49.8419, 24.0315];
    const xZoom = zoom || 16;
    const style = {
        width: '100%',
        height: '100%',
    };

    useEffect(() => {
        mapRef.current = L.map(sid, {
            center: xCenter,
            zoom: xZoom,
            layers: [
                L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {
                    attribution:
                        '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
                }),
            ],
        });
    }, [xCenter, xZoom, sid]);

    // add layer
    const layerRef = useRef(null);
    useEffect(() => {
        layerRef.current = L.layerGroup().addTo(mapRef.current);
    }, []);

    // update markers
    useEffect(
        () => {
            layerRef.current.clearLayers();
            markersData.forEach((marker) => {
                L.marker(marker.latLng, { title: marker.title }).addTo(
                    layerRef.current,
                );
            });
        },
        [markersData],
    );

    return <div ref={mapRef} id={sid} style={style} />;
};

CustomMap.propTypes = {
    markersData: PropTypes.arrayOf(
        PropTypes.shape({
            title: PropTypes.string.isRequired,
            latLng: PropTypes.shape({
                lat: PropTypes.number.isRequired,
                lng: PropTypes.number.isRequired,
            }).isRequired,
        }),
    ),
    center: PropTypes.arrayOf(
        PropTypes.number,
    ),
    zoom: PropTypes.number,
};

CustomMap.defaultProps = {
    markersData: [],
    zoom: 12,
    center: [49.8419, 24.0315],
};


export default CustomMap;
