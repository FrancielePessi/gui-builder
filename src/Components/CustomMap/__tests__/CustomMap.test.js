import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import CustomMap from '../CustomMap';

configure({ adapter: new Adapter() });

describe('CustomMap', () => {
    it('should render correctly setting the data', () => {
        const props = {
            markersData: [{
                title: 'Point A',
                latLng: {
                    lat: 40,
                    lng: 35,
                },
            }],
            center: [39.8419, 34.0315],
        };
        const wrapper = shallow(<CustomMap {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
