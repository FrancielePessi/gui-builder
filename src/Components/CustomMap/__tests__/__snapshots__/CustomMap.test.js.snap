// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomMap check type props 1`] = `
<div
  id="149583175659"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;

exports[`CustomMap should render correctly setting the data 1`] = `
<div
  id="429428167371"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;
